demos/object_detection_demo_ssd_async/main.cpp

```diff
@@ -354,58 +403,194 @@ int main(int argc, char *argv[]) {
+                        modeInfo[isUserSpecifiedMode].lastStartTime = switchTime;
+                    } else {
+                        presenter.handleKey(key);
+                    }
+                }
+            } else if (!emptyRequests.empty() && cap.isOpened()) {
+                auto startTime = std::chrono::steady_clock::now();
+                
+                cv::Mat frame;
+                if (!cap.read(frame)) {
+                    if (frame.empty()) {
+                        if (FLAGS_loop_input) {
+                            cap.open((FLAGS_i == "cam") ? 0 : FLAGS_i.c_str());
+                        } else cap.release();
+                        continue;
+                    } else {
+                        throw std::logic_error("Failed to get frame from cv::VideoCapture");
+                    }
+                }
 
-            if (isModeChanged) {
-                isModeChanged = false;
-            }
+                InferRequest::Ptr request;
+                {
+                    std::lock_guard<std::mutex> lock(mutex);
 
-            // Final point:
-            // in the truly Async mode we swap the NEXT and CURRENT requests for the next iteration
-            curr_frame = next_frame;
-            next_frame = cv::Mat();
-            if (isAsyncMode) {
-                async_infer_request_curr.swap(async_infer_request_next);
-            }
+                    request = std::move(emptyRequests.front());
+                    emptyRequests.pop_front();
+                }
+                frameToBlob(frame, request, imageInputName);
+                
+                bool frameMode = isUserSpecifiedMode;
+                request->SetCompletionCallback([&mutex,
+                                                &completedRequestResults,
+                                                nextFrameId,
+                                                frame,
+                                                request,
+                                                &outputName,
+                                                startTime,
+                                                frameMode,
+                                                &isUserSpecifiedMode,
+                                                &emptyRequests,
+                                                &callbackException,
+                                                &condVar] {
+                    {
+                        std::lock_guard<std::mutex> callbackLock(mutex);
+                    
+                        try {
+                            completedRequestResults.insert(
+                                std::pair<int, RequestResult>(nextFrameId, RequestResult{
+                                    frame,
+                                    std::unique_ptr<LockedMemory<const void>>(new
+                                        LockedMemory<const void>(as<MemoryBlob>(request->GetBlob(outputName))->rmap())),
```

Comment:
Why do you map the blob here? IMO, it would be a lot clearer to do it where the contents of the blob are actually used.

Reply:
Done